website/api_versioned_docs/version-0.19/bonsai/bonsai-on-eth.md

```diff
@@ -2,18 +2,133 @@
+Bonsai can be used as a [zk coprocessor] for Ethereum, allowing you to request [verified proofs] via an off-chain REST API interface or on-chain directly from your smart contract.
 
 At a high level, here's how it works:
 
-![Bonsai ETH Relay overview](/img/eth-relay-diagram.jpg)
+![Bonsai ETH Relay overview](/img/bonsai_ethereum.png)
 
-<!-- TODO: Highlight the relay sections of this diagram -->
+1. Users can delegate their smart contract's logic to Bonsai via a `Request Callback` interface. This interface is accessible both _off-chain_ through the HTTP REST API of the `Bonsai Relayer` and _on-chain_ via the [Bonsai Relay Contract].
+2. The `Bonsai Relayer` sends the proof request to Bonsai.
+3. Bonsai generates a [Groth16 SNARK proof] and its result, encapsulated in a journal.
+4. The `Bonsai Relayer` submits this proof and journal on-chain to the [Bonsai Relay Contract] for validation.
+5. If validated, the journal is dispatched to the user's smart contract via the specified callback.
 
-Our Bonsai-ETH Relay acts as a middle-man between your app contract and the Bonsai proving service. In order to use the Bonsai ETH Relay, dApp developers must have completed the following steps:
+To integrate your application with Bonsai, you'll need to:
 
-1. Deployed a [smart contract] on chain that calls the relay contract
-2. Uploaded a compiled [program] for the zkVM to the Bonsai proving service
+1. Write the [program] you want proven and upload it to Bonsai
+2. Write the on-chain part of your application and deploy it
+
+## Interfaces
+
+The Bonsai Ethereum Relay provides both an _off-chain_ (vie the HTTP REST API of the `Bonsai Relayer`) and an _on-chain_ (via the [Bonsai Relay Contract]) interface to send `Callback requests`.
+
+### Off-chain
+
+A typical flow works as follows:
+
+1. Deploy the `Bonsai Relay Contract` to Ethereum using address `0xB..`.
+2. Launch the `Bonsai Relayer`, setting `--contract-address` to `0xB..`.
+3. To delegate off-chain computation for Smart Contract `A` to Bonsai, register its `Image` or `ELF` (the binary executing the computation on RISC Zero ZKVM) with Bonsai.
+4. Use the generated `Image ID` and the `Bonsai Relay Contract` address 0xB.. to deploy Smart Contract `A` on Ethereum.
+5. Initiate a `Callback request` through the `Bonsai Relayer`'s off-chain REST API. This request is sent to Bonsai.
+6. After Bonsai produces an execution proof, the `Bonsai Relayer` relays the proof and computation result to the `Bonsai Relay Contract`.
+7. Ethereum verifies the proof on-chain. If validated, Smart Contract `A` receives the computation result via the `invoke_callback` function.
+
+#### Example
+
+The following example assumes that the `Bonsai Relayer` is up and running with the server API enabled,
+and that the memory image of your `ELF` is already registered against Bonsai with a given `IMAGE_ID` as its identifier.
+
+```rust,no_run
```

Comment:
It is awkward from a maintenance standpoint to have this much code and writing duplicated between the README and the website. If you're onboard from doing maintenance I guess it's ok, but de-duplicating docs would be nice if it feels reasonable to have one reference the other.